algorithms/src/Kokkos_Algo.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_ALGO_HPP
+#define KOKKOS_ALGO_HPP
+
+#include <Kokkos_Core.hpp>
+
+/// \file Kokkos_Algo.hpp
+/// \brief Kokkos counterparts for Standard C++ Library algorithms
+
+namespace Kokkos {
+namespace Experimental {
+
+template <class DataType, class... Properties>
+auto begin(Kokkos::View<DataType, Properties...>& v) -> decltype(v.data()) {
```

Comment:
what do you mean by "layout is unique"?

Reply:
not sure about the unique part, but IIRC we have agreed to limit this to `LayoutLeft`, `LayoutRight` and `LayoutStride`

Reply:
if you have a unique contiguous layout that would be fine too, but I am ok to restrict to LayoutLeft/Right/Stride for now. Ah just noticed we don't have that trait in Kokkos yet, its only in mdspan. So technically we allow layouts to be non-unique. For example you could have a Symmetric 2D layout which return for mapping(i,k) = i>k ? i*N+k : k*N+i; So (i,k) returns the same index as (k,j). Now if want to sum up all values of the NxN matrix you need to loop over the entire index range NxN, but if you want to just multiply each element by 2 you need to be careful to just iterate over the upper or lower triangle so you don't multiply everything other than the diagonal by 4. In mdspan we have a trait which tells you whether the layout is unique, so you can at least reason about this. Since we don't have that yet in Kokkos restricting to LayoutLeft/RIght/Stride is the right thing to do.

Reply:
yeah that is why I was confused, I did not see any unique layout trait!